sessiontxn/isolation/readcommitted.go

```diff
@@ -230,14 +284,52 @@ func (p *PessimisticRCTxnContextProvider) AdviseWarmup() error {
 	return nil
 }
 
+func planSkipGetTsoFromPD(sctx sessionctx.Context, plan plannercore.Plan, inLockOrWriteStmt bool) bool {
+	switch v := plan.(type) {
+	case *plannercore.PointGetPlan:
+		return sctx.GetSessionVars().RcPointLockReadUseLastTso && (v.Lock || inLockOrWriteStmt)
+	case plannercore.PhysicalPlan:
+		if len(v.Children()) == 0 {
+			return false
+		}
+		_, isPhysicalLock := v.(*plannercore.PhysicalLock)
+		for _, p := range v.Children() {
+			if !planSkipGetTsoFromPD(sctx, p, isPhysicalLock || inLockOrWriteStmt) {
+				return false
+			}
+		}
+		return true
+	case *plannercore.Update:
+		return planSkipGetTsoFromPD(sctx, v.SelectPlan, true)
+	case *plannercore.Delete:
+		return planSkipGetTsoFromPD(sctx, v.SelectPlan, true)
+	case *plannercore.Insert:
+		// RcInsertUseLastTso controls whether to make a tso request, but not to whether to wait tso requst.
+		// Insert sqls are always optimized to use last tso here.
+		return v.SelectPlan == nil
+	}
+	return false
+}
+
 // AdviseOptimizeWithPlan in RC covers much fewer cases compared with pessimistic repeatable read.
-// We only optimize with insert operator with no selection in that we do not fetch latest ts immediately.
-// We only update ts if write conflict is incurred.
+// We do not fetch latest ts immediately for such scenes.
+// 1. PointGet of SELECT ... FOR UPDATE  2. Insert without select 3. update by PointGet 4. delete by PointGet.
+// The func `planner.Optimize` always calls the `txnManger.AdviseWarmup` to warmup in the past, it makes a tso request
+// for all sqls execept `SELECT` with RcCheckTs, but whether to use the tso request and wait it are depended
+// on `AdviseOptimizeWithPlan`.
+// (a) text protocol request
+// For insert statements without select, they always don't use the above tso, don't wait for it either.
+// If tidb_rc_insert_use_last_tso is ON, we don't make tso requests too, we use the p.latestOracleTS.
+// If tidb_rc_insert_use_last_tso is OFF, we make tso requests but don't wait.
+// For update/delete/select statements, if tidb_rc_point_lock_read_use_last_tso is ON, we don't make tso requests,
+// we use the p.latestOracleTS.
+// (b) binary protocol request(prepare mode)
+// The behaviour is as text protocol request at most time, except that the it can get a plan from plan cache,
+// when optimizer doesn't call txnManger.AdviseWarmup forever.
```

Comment:
There're some grammar mistakes, you could use something like `Grammarly` to do some checks.

Reply:
I have checked in MS-Word, and fixed it.